### DISCUSSION TITLE
Temperature sensor widely varies

### DISCUSSION THREAD

**Discussion by Geo-Ron** [jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip#40] [Q&A]
@jazzmonger 

Sorry to reach out to you again. I hope you are willing to help.

When we first installed the stove, it indicated a room temperature of 84 ºC (184ºF) which .

I found out that the sensor was soldered on. I measured it and found out it was a 10KΩ NTC thermistor.
The resistance of the thermistor roughly correlated to a 100KΩ B3950 NTC thermistor, so I swapped the component.

I added a 100nF capacitor in parallel to smooth the measurements.

This improved the measurements by the stove a lot!

The only thing is, it constantly fluctuates between 17ºC and 26ºC now with a room temperature of about 20ºC.

Have you got any clue, measurements or characteristics about the original room temperature sensor?

**Comment by jazzmonger** [comment]
I'm always willing to help!  I share your frustation!

I think I have one in my spare parts bin.  I’m happy to measure it.  I actually hate thermistors. their resistance can vary WIDELY at constant temps.  The ones in the pellet stoves are particularly bad.... Here's my graph of a "constant" temp of my shop pellet stove.  The smooth line is my ESP DHT11 sensor 16.5C-17.22C (0.72C) and is the REAL temp in my shop.  The choppy line is the thermistor room sensor in my pellet stove. it varies from 15C to 20C (5C) while yours varies from 17-26C (9C).

![image](https://github.com/user-attachments/assets/044192f1-b0c5-44d3-9f1b-b3233c2e7343)

I thought about trying to make them behave better but never got a chance - have you played with a circuit like this? 3 caps/2 resistors.

https://www.doeeet.com/content/eee-components/passives/thermistor-based-temperature-sensing-system-explained/

EDIT - I actually had 2 spare OEM sensors.  I hooked one to my trusty & rock solid Fluke Multimeter and watched the resistance vary from 144.2k - 145.6k over about 5 min at a constant room temp of 17.11C.  thats 1400 ohms. no wonder these things are all over the place.  They both did the same thing.

**Reply by jazzmonger** [reply]
Interestingly one of the OEMS of this control unit figured out this problem and created a different design:

![image](https://github.com/user-attachments/assets/56e3c05c-2881-4899-bd4d-6ffa4e5fbfa3)

Notice the DHT11 room sensor located IN the top of the control unit, not inside the stove.  I'm sure it does a MUCH better job of regulating (or is it just for reporting? unsure) the room temp. Unfortunately (depending on how you look at it) it doesn't run the Tuya code in the main MCU so we couldn't get it working w/ my mods.

In practice I've noticed that the firmware inside our units takes this thermocouple sensor noise somewhat into account.  The room temp has to be above the set temp for a couple of minutes before the stove turns on (ECO1) or goes from Idle to Stabilization (ECO2), providing some level of hysteresis control that seems to work ok.  Your sensor swing of 9C might prove challenging.

**Comment by Geo-Ron** [comment]
What I am thinking about is making my esp control the resistance on these pins. 
One resistance value for making the room temperature about 85 degrees celsius, so the unit always thinks it is warm enough. The other for making the room temperature around 0 degrees celsius, so the unit will need to warm the room.

Combine this with the set room temperature and the measurements of a ds18b20 sensor  will make a great control over what the unit is doing. 

The difficulty is, what happens **around** the set temperature... a pid controller perhaps.

**Reply by jazzmonger** [reply]
Interesting idea. Controlling the resistance might work - I hadn’t thought of that!

I built an ESPHome based PID controller for the pad heaters on my solar battery banks. It works really well. 
some good info, code and insight on it here:

https://diysolarforum.com/threads/new-battery-heater-controller-home-assistant-and-esphome.53665

**Comment by Geo-Ron** [comment]
By the way, it is not the thermistor to blame.
I tested is with an esp32 using https://github.com/Geo-Ron/esphome-PelletStove/blob/main/code/ntc-test.ino 

The measured temperature swing was about 0.5 degrees. I did not bother checking the relation to the actual temperature.

Thanks for the reference to https://esphome.io/components/climate/pid.html

**Reply by jazzmonger** [reply]
> By the way, it is not the thermistor to blame.

what?  its more shoddy code ? in this Pellet stove?  I'm shocked :)

**Comment by Geo-Ron** [comment]
Hey @jazzmonger ,

Have you got any clue on the connector used in these sensors?

The connector that comes closest, it the [jst xap-02v1](https://www.digikey.com/en/products/detail/jst-sales-america-inc/xap-20v-1/923774). But this is not the correct one...

I measure the pitch to be 2.54mm


![image](https://github.com/user-attachments/assets/2586f55c-dec4-4aa1-8510-e20a41662e9f)

**Comment by jazzmonger** [comment]
VHR-2 JST
https://www.amazon.com/Kidisoii-VH3-96mm-Connector-Pre-Crimped-2Pin-20CM/dp/B0C3LYBSK3?th=1

**Reply by Geo-Ron** [reply]
Are you sure? According to [digikey](https://www.digikey.com/en/products/detail/jst-sales-america-inc/VHR-2M/9924200) these have a 3.96mm pitch. And mine have a 2.54mm pitch...

**Comment by jazzmonger** [comment]
Hmmm... ok, let me check - I bought some from Alibaba for this exact purpose.

Heres the exact product I bought and I can confirm they fit.
https://www.aliexpress.us/item/2255799934230488.html?spm=a2g0o.order_list.order_list_main.4.10651802jopgM4&gatewayAdapt=glo2usa
